### PR TITLE
Update repository to match module template (February 2024)

### PR DESCRIPTION
**Pull request by Mrtenz**
This updates the repository to match the module template as of February 2024. This repo hasn't been updated in a while, so there's quite a few changes. The most notable changes are:

- The GitHub workflows were updated to use the latest versions of the publishing actions, including a Slack message for publishing.
- Some actions now only run on the `.nvmrc` Node versions, to reduce the number of jobs.
- The Yarn allow-scripts plugin was updated, and the constraints plugin was added.
- The package is now built with `tsup`, and includes a ESM build.

### CODE REVIEW

**Comment by Mrtenz** [comment]
@SocketSecurity ignore npm/debug@2.6.9
@SocketSecurity ignore npm/agent-base@7.1.0
@SocketSecurity ignore npm/mz@2.7.0
@SocketSecurity ignore npm/@npmcli/agent@2.2.1
@SocketSecurity ignore npm/socks-proxy-agent@8.0.2

Network access is ok. Not used in production.

@SocketSecurity ignore npm/validate-npm-package-name@5.0.0
@SocketSecurity ignore npm/nopt@7.2.0
@SocketSecurity ignore npm/abbrev@2.0.0
@SocketSecurity ignore npm/source-map@0.8.0-beta.0
@SocketSecurity ignore npm/normalize-package-data@6.0.0

New authors seem ok.

@SocketSecurity ignore npm/foreground-child@3.1.1
@SocketSecurity ignore npm/tree-kill@1.2.2

Shell access ok. Not used in production.

@SocketSecurity ignore npm/esbuild@0.19.12

Network / shell access is expected.

@SocketSecurity ignore npm/tsup@7.3.0

Deprecated is ok. Shouldn't be an issue in our case.

@SocketSecurity ignore npm/@lavamoat/preinstall-always-fail@2.0.0

Script is expected.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1,4 +1,4 @@
 # Lines starting with '#' are comments.
 # Each line is a file pattern followed by one or more owners.
 
-*                                    @MetaMask/devs
+*                                    @MetaMask/engineering
```
Comment (FrederikBolding) [inline .github/CODEOWNERS:4]: GitHub seems to think this doesn't exist. Does the group need to be added manually to the repo or something?

Comment (Mrtenz) [reply .github/CODEOWNERS:4]: Yeah. I already asked @Gudahtt to do so.

**Review by FrederikBolding — COMMENTED** [review]
LGTM once the CODEOWNERS change has been resolved

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**